python/nvtx/utils/nop.py

```diff
@@ -0,0 +1,27 @@
+# Copyright 2020  NVIDIA Corporation.  All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Utilities for implementing no-op functions/methods
+
+import contextlib
+
+from functools import wraps
+
+
+class NullContextDecorator(contextlib.nullcontext):
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def __call__(self, func):
+        return func
+
+
+def disable(func):
+    # disable the wrapped function
+    @wraps(func)
+    def inner(*args, **kwargs):
+        pass
+    return inner
```

Comment:
This still incurs the overhead of having an inner function call, right? Also, handling `*args, **kwargs` is also surprisingly not free.

What about if we read the environment variable once at library load time and then defined the functions to only `pass` if the environment variable is set?

Reply:
You're right: using a decorator like this incurs a significant overhead compared with afunction that just does `pass`:

```python
In [48]: def foo(x, y):
    ...:     pass
    ...:

In [49]: %timeit foo(1, 2)
121 ns ± 0.789 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

```python
In [51]: @disable
    ...: def foo(x, y):
    ...:     return x + y
    ...:
    ...:

In [52]: %timeit foo(1, 2)
198 ns ± 1.95 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Barring the use of a decorator though, our options are:

1. Write duplicate functions with the same signature that `pass`
2.  "Programatically" define functions with `exec`. Something similar was suggested  here: https://stackoverflow.com/a/25460056

Using an [`exec` based solution](https://gist.github.com/shwina/cd515c05d39b2618d1c8563f0507377d) gets us where we want:

```python
In [12]: @disable
    ...: def foo(x, y=1):
    ...:     pass
    ...:

In [13]: foo
Out[13]: <function foo(x, y=1)>

In [14]: foo(1, 2)

In [15]: %timeit foo(1, 2)
133 ns ± 0.909 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

Reply:
Discussed offline, we decided  to go with (1) for now and steer clear of `exec`.

In the future, we  might consider a preprocessing step to autogenerate the no-op functions.